File: src/main/java/com/chessapp/api/game/BoardPosition.java
package com.chessapp.api.game;

import com.chessapp.api.pieces.piece.ChessPiece;
import com.chessapp.api.pieces.utils.InvalidPositionException;

/**
 * The board position manager, used to manage
 * the positions of all of the Chess pieces
 */
public class BoardPosition {

    // The internal array used to keep track of the location of pieces
    private final ChessPiece[][] positions;

    // The size of the board
    private final int boardSize;

    /**
     * Constructor for creating a board position manager
     * with a particular size
     *
     * @param _boardSize The desired size of the board (N x N)
     */
    public BoardPosition(int _boardSize) {
        assert (_boardSize >= 8);
        positions = new ChessPiece[_boardSize][_boardSize];
        boardSize = _boardSize;
    }

    /**
     * Default constructor for board manager, will default
     * to a size of N = 8
     */
    public BoardPosition() {
        positions = new ChessPiece[8][8];
        boardSize = 8;
    }

    /**
     * Determines if a particular position is occupied by a piece
     *
     * @param xPos The x-position of the piece
     * @param yPos The y-position of the piece
     * @return Whether or not the position is occupied
     */
    public Boolean isOccupied(int xPos, int yPos) {
        return positions[xPos][yPos] != null;
    }

    /**
     * Occupies a position in the position manager
     *
     * @param xPos  The desired x-position in the manager
     * @param yPos  The desired y-position in the manager
     * @param piece The desired piece to occupy the position
     * @throws InvalidPositionException thrown if the position is currently occupied
     */
    public void occupyPosition(int xPos, int yPos, ChessPiece piece) throws InvalidPositionException {
        if (isOccupied(xPos, yPos)) {
            throw new InvalidPositionException("Position is currently occupied by a piece");
        } else {
            positions[xPos][yPos] = piece;
        }
    }

    /**
     * Allows a piece to leave the position
     *
     * @param piece The desired piece to leave
     */
    public void leavePosition(ChessPiece piece) {
        positions[piece.x()][piece.y()] = null;
    }

    /**
     * Public getter for the size of the board
     *
     * @return The size of the board
     */
    public int getSize() {
        return boardSize;
    }

    /**
     * Public getter for getting piece at particular position
     *
     * @param xPos X-position we are reading from
     * @param yPos Y-position we are reading from
     * @return The piece at position (x,y) in the board
     */
    public ChessPiece getPieceAtPosition(int xPos, int yPos) {
        return positions[xPos][yPos];
    }
}

File: src/main/kotlin/com/chessapp/api/board/BoardPosition.kt
package com.chessapp.api.board

import com.chessapp.BOARD_SIZE
import com.chessapp.api.pieces.piece.ChessPiece
import com.chessapp.api.pieces.piece.PieceColor
import com.chessapp.api.pieces.utils.InvalidPositionException

/**
 * The board position manager, used to manage
 * the positions of all the Chess pieces
 */
class BoardPosition {

    // The internal array used to keep track of the location of pieces
    private val positions: Array<Array<ChessPiece?>> = Array(BOARD_SIZE) { arrayOfNulls(BOARD_SIZE) }

    fun getPieceAtPosition(rank: File, file: Int): ChessPiece? {
        val (x, y) = PositionUtils.getCoordinatesFromFileRank(rank, file)
        return positions[x][y]
    }

    fun isPositionOccupiedByPiece(rank: File, file: Int): Boolean {
        val (x, y) = PositionUtils.getCoordinatesFromFileRank(rank, file)
        return positions[x][y] != null
    }

    fun occupyPosition(piece: ChessPiece) {
        val rank = piece.rank()
        val file = piece.file()
        occupyPosition(file, rank, piece)
    }

    fun movePiece(piece: ChessPiece, newFile: File, newRank: Int) {
        val gotPiece = getPieceAtPosition(piece.file(), piece.rank())
        if (gotPiece != piece) {
            throw InvalidPositionException("Piece internal coordinates do not line up with board position: $piece is has position ${piece.file()}${piece.rank()}, " +
                    "whereas piece at this position on the board is $gotPiece")
        }
        if (isPositionOccupiedByPiece(newFile, newRank)) {
            // TODO update this to handle captures etc
            throw InvalidPositionException("Square $newFile$newRank is already occupied by another piece")
        }

        removePiece(piece)
        occupyPosition(newFile, newRank, piece)
        piece.setPos(newFile, newRank)
    }

    private fun occupyPosition(file: File, rank: Int, piece: ChessPiece) {
        if (isPositionOccupiedByPiece(file, rank)) {
            throw InvalidPositionException("Square $file$rank is already occupied by another piece")
        }

        val (x, y) = PositionUtils.getCoordinatesFromFileRank(file, rank)
        positions[x][y] = piece
    }

    fun removePiece(piece: ChessPiece): ChessPiece? = removePiece(piece.file(), piece.rank())

    private fun removePiece(rank: File, file: Int): ChessPiece? {
        val (x, y) = PositionUtils.getCoordinatesFromFileRank(rank, file)
        return positions[x][y].also { positions[x][y] = null }
    }

    override fun toString(): String {
        val size = BOARD_SIZE
        val rotatedArray = Array(size) { Array<ChessPiece?>(size) { null } }

        // Rotating the array 90 degrees counterclockwise
        for (i in 0 until size) {
            for (j in 0 until size) {
                rotatedArray[size - j - 1][i] = positions[i][j]
            }
        }

        // Column labels (A to H)
        val columnLabels = ('A' until 'A' + size).joinToString(" ") { it.toString() }

        // Building the board with row numbers (8 to 1) and columns (A to H)
        return buildString {
            append("  ") // Leading spaces for row number alignment
            append(columnLabels)
            appendLine()

            for (i in 0 until size) {
                append(8 - i) // Row numbers (8 to 1)
                append(" ") // Space between row number and the row itself

                append(rotatedArray[i].joinToString(" ") { piece ->
                    piece?.let {
                        if (it.color() == PieceColor.WHITE) it.toString().uppercase() else it.toString().lowercase()
                    } ?: "."
                })

                append(" ") // Space between the row and the row number
                append(BOARD_SIZE - i) // Repeat row number at the end of the row
                appendLine()
            }

            append("  ") // Leading spaces for alignment
            append(columnLabels) // Bottom column labels
        }
    }
}
